src/transports/ssh.c

```diff
@@ -638,6 +639,15 @@ static int _git_ssh_setup_conn(
 		}
 
 		error = _git_ssh_authenticate_session(session, cred);
+
+		if (error == GIT_EAUTH) {
```

Comment:
You can clobber `error` here, so no need to declare `sub_error`.

```c
/* refresh auth methods */
if (error == GIT_EAUTH &&
    error = list_auth_methods(…))
        goto done;
```

Reply:
I disagree, with that change `error` will no longer be `GIT_EAUTH` after the call to `list_auth_methods()` and cancel the loop.

Reply:
Actually… there's nothing in here that actually prevents a server from swamping you into a loop by repeatedly switching auth methods and continuously failing to authenticate, so maybe that's something that needs to happen… Not really related, but you've just mentioned a loop, and I happened to completely miss its condition 😉.

So; I'm okay with your initial version then. If you happen to add a auth counter or something, that would be lovely. I'm not sure how it would *really work* though…

Reply:
> So; I'm okay with your initial version then. If you happen to add a auth counter or something, that would be lovely. I'm not sure how it would really work though…

Does the server even need to switch auth methods? He could just refuse the authentication repeatedly, which would cause us to loop infinitely here until one manages to send the correct credentials. Or am I missing something?

Reply:
@pks-t Authentication can be cancelled by the callback by returning an error code